File: app/src/main/java/com/bluzi/babymonitor/data/Store.kt
package com.bluzi.babymonitor.data

import com.bluzi.babymonitor.log.Log
import com.bluzi.babymonitor.xiaomi.Device
import com.bluzi.babymonitor.xiaomi.Session
import com.bluzi.babymonitor.xiaomi.base64ToBytes
import com.bluzi.babymonitor.xiaomi.toBase64
import org.json.JSONObject

// Persistence: session (encrypted — AUTH-5/6), selected camera (CAM-2/3), settings
// (LIVE-2, ALRM-6). Pure serialization + seams; Android impls live in AndroidStore.kt.

interface KeyValueStore {
    fun get(key: String): String?
    fun put(key: String, value: String)
    fun remove(key: String)
}

/** Encryption seam for secrets — Android Keystore in production, passthrough fake in tests. */
interface SecretBox {
    fun seal(plain: String): String
    fun open(sealed: String): String?
}

object Codecs {
    fun sessionToJson(s: Session): String = JSONObject()
        .put("userId", s.userId)
        .put("cUserId", s.cUserId)
        .put("passToken", s.passToken)
        .put("serviceToken", s.serviceToken)
        .put("ssecurity", s.ssecurity.toBase64())
        .put("region", s.region)
        .toString()

    fun sessionFromJson(json: String): Session? = try {
        val v = JSONObject(json)
        Session(
            userId = v.getString("userId"),
            cUserId = v.optString("cUserId"),
            passToken = v.getString("passToken"),
            serviceToken = v.getString("serviceToken"),
            ssecurity = v.getString("ssecurity").base64ToBytes(),
            region = v.getString("region"),
        )
    } catch (_: Exception) {
        null
    }

    fun deviceToJson(d: Device): String = JSONObject()
        .put("did", d.did)
        .put("name", d.name)
        .put("model", d.model)
        .put("mac", d.mac)
        .put("ip", d.ip)
        .toString()

    fun deviceFromJson(json: String): Device? = try {
        val v = JSONObject(json)
        Device(
            did = v.getString("did"),
            name = v.optString("name"),
            model = v.optString("model"),
            mac = v.optString("mac"),
            ip = v.optString("ip"),
        )
    } catch (_: Exception) {
        null
    }
}

data class Settings(
    val muted: Boolean = false,
    val alarmEnabled: Boolean = false,
    // ALRM-2: the one detection control — 1 (needs loud crying) .. 10 (quiet crying), middle default.
    val alarmSensitivity: Int = SENSITIVITY_DEFAULT,
    // ALRM-7: "always" or a daily window (minutes since midnight, may cross midnight).
    val alarmScheduleMode: String = SCHEDULE_ALWAYS,
    val alarmWindowStartMinutes: Int = 19 * 60,
    val alarmWindowEndMinutes: Int = 7 * 60,
    // WATCH-1: feed watchdog.
    val watchdogEnabled: Boolean = false,
    val watchdogGraceSeconds: Int = 30,
    // ALRM-11: how each alarm sounds. The two must never be the same (see [withSounds]).
    val cryAlarmSound: String = SOUND_RISING_CHIME,
    val feedAlarmSound: String = SOUND_LOW_PULSE,
    val alarmVolume: Double = 0.85, // 0..1, on top of the phone's alarm stream
    val alarmVibrate: Boolean = true,
) {
    /**
     * ALRM-11: the two alarms must never sound the same — "the baby is crying" and "the monitor is
     * blind" call for different reactions, and at 3am you decide by ear before you can read.
     * Picking a sound for one alarm therefore pushes the other off it.
     */
    fun withSounds(cry: String = cryAlarmSound, feed: String = feedAlarmSound): Settings {
        if (cry != feed) return copy(cryAlarmSound = cry, feedAlarmSound = feed)
        val cryChanged = cry != cryAlarmSound
        val alternative = ALARM_SOUNDS.first { it != cry }
        return if (cryChanged) {
            copy(cryAlarmSound = cry, feedAlarmSound = alternative)
        } else {
            copy(cryAlarmSound = alternative, feedAlarmSound = feed)
        }
    }

    fun toJson(): String = JSONObject()
        .put("muted", muted)
        .put("alarmEnabled", alarmEnabled)
        .put("alarmSensitivity", alarmSensitivity)
        .put("alarmScheduleMode", alarmScheduleMode)
        .put("alarmWindowStartMinutes", alarmWindowStartMinutes)
        .put("alarmWindowEndMinutes", alarmWindowEndMinutes)
        .put("watchdogEnabled", watchdogEnabled)
        .put("watchdogGraceSeconds", watchdogGraceSeconds)
        .put("cryAlarmSound", cryAlarmSound)
        .put("feedAlarmSound", feedAlarmSound)
        .put("alarmVolume", alarmVolume)
        .put("alarmVibrate", alarmVibrate)
        .toString()

    companion object {
        const val SCHEDULE_ALWAYS = "always"
        const val SCHEDULE_WINDOW = "window"

        // ALRM-2: the sensitivity scale. 1 needs loud crying, 10 alarms on quiet crying.
        const val SENSITIVITY_MIN = 1
        const val SENSITIVITY_MAX = 10
        const val SENSITIVITY_DEFAULT = 5

        // ALRM-4: never all the way down — a stored volume of zero would be a silent alarm.
        const val VOLUME_MIN = 0.2
        const val VOLUME_MAX = 1.0

        // WATCH-1: grace bounds — a zero grace would alarm on every one-second audio hiccup.
        const val GRACE_MIN_SECONDS = 5
        const val GRACE_MAX_SECONDS = 120

        // ALRM-11: calm → urgent.
        const val SOUND_SOFT_CHIME = "soft-chime"
        const val SOUND_RISING_CHIME = "rising-chime"
        const val SOUND_LOW_PULSE = "low-pulse"
        const val SOUND_URGENT_BEEP = "urgent-beep"
        const val SOUND_SIREN = "siren"

        val ALARM_SOUNDS = listOf(
            SOUND_SOFT_CHIME,
            SOUND_RISING_CHIME,
            SOUND_LOW_PULSE,
            SOUND_URGENT_BEEP,
            SOUND_SIREN,
        )

        fun fromJson(json: String?): Settings {
            if (json == null) return Settings()
            return try {
                val v = JSONObject(json)
                val settings = Settings(
                    muted = v.optBoolean("muted", false),
                    alarmEnabled = v.optBoolean("alarmEnabled", false),
                    alarmSensitivity = v.optInt("alarmSensitivity", legacySensitivity(v))
                        .coerceIn(SENSITIVITY_MIN, SENSITIVITY_MAX),
                    alarmScheduleMode = v.optString("alarmScheduleMode", SCHEDULE_ALWAYS),
                    alarmWindowStartMinutes = v.optInt("alarmWindowStartMinutes", 19 * 60)
                        .coerceIn(0, 24 * 60 - 1),
                    alarmWindowEndMinutes = v.optInt("alarmWindowEndMinutes", 7 * 60)
                        .coerceIn(0, 24 * 60 - 1),
                    watchdogEnabled = v.optBoolean("watchdogEnabled", false),
                    // Clamped like sensitivity: stored data (old versions, hand edits) must never
                    // yield a watchdog that fires on every hiccup or an alarm that plays silently.
                    watchdogGraceSeconds = v.optInt("watchdogGraceSeconds", 30)
                        .coerceIn(GRACE_MIN_SECONDS, GRACE_MAX_SECONDS),
                    cryAlarmSound = v.optString("cryAlarmSound", SOUND_RISING_CHIME),
                    feedAlarmSound = v.optString("feedAlarmSound", SOUND_LOW_PULSE),
                    alarmVolume = v.optDouble("alarmVolume", 0.85).coerceIn(VOLUME_MIN, VOLUME_MAX),
                    alarmVibrate = v.optBoolean("alarmVibrate", true),
                )
                // Stored data can be old or hand-edited; never let the two alarms collide (ALRM-11).
                settings.withSounds()
            } catch (_: Exception) {
                Settings()
            }
        }

        /**
         * ALRM-6: settings written before the sensitivity slider carried a dB threshold
         * (3 = most sensitive .. 24 = least). Map it onto the 1..10 scale rather than lose the
         * user's tuning. The old scale is frozen history — these constants must never track
         * whatever the live sensitivity mapping becomes.
         */
        private fun legacySensitivity(v: JSONObject): Int {
            if (!v.has("alarmThresholdDb")) return SENSITIVITY_DEFAULT
            return Math.round((24.0 - v.optDouble("alarmThresholdDb", 14.0)) / 2.0).toInt()
                .coerceIn(SENSITIVITY_MIN, SENSITIVITY_MAX)
        }
    }
}

class AppStore(private val kv: KeyValueStore, private val box: SecretBox) {
    private companion object {
        const val KEY_SESSION = "session_v1"
        const val KEY_DEVICE = "device_v1"
        const val KEY_SETTINGS = "settings_v1"
        const val KEY_MONITORING = "monitoring_v1"
        const val KEY_ALARM_VOLUME_PREV = "alarm_volume_prev_v1"
        const val KEY_ALARM_VOLUME_RAISED = "alarm_volume_raised_v1"
        const val KEY_CRY_CALIBRATION = "cry_calibration_v1" // JSON object: camera did → steps
    }

    /**
     * AUTH-6: tokens are only ever stored encrypted. If the Keystore refuses (key invalidated by
     * an OS update or a lock-screen change), drop the session rather than crash or, worse, fall
     * back to plaintext — the user simply signs in again.
     */
    fun saveSession(s: Session) {
        try {
            kv.put(KEY_SESSION, box.seal(Codecs.sessionToJson(s)))
        } catch (e: Exception) {
            Log.w("data", "could not encrypt session, not persisting it: ${e.message}", e)
            kv.remove(KEY_SESSION)
        }
    }

    fun loadSession(): Session? =
        kv.get(KEY_SESSION)?.let { box.open(it) }?.let { Codecs.sessionFromJson(it) }

    fun saveDevice(d: Device) = kv.put(KEY_DEVICE, Codecs.deviceToJson(d))

    fun loadDevice(): Device? = kv.get(KEY_DEVICE)?.let { Codecs.deviceFromJson(it) }

    fun loadSettings(): Settings = Settings.fromJson(kv.get(KEY_SETTINGS))

    fun saveSettings(s: Settings) = kv.put(KEY_SETTINGS, s.toJson())

    /** BG-10: was monitoring running when the process last went away (e.g. a reboot)? */
    fun wasMonitoring(): Boolean = kv.get(KEY_MONITORING) == "true"

    fun setMonitoring(on: Boolean) = kv.put(KEY_MONITORING, on.toString())

    /**
     * ALRM-10: the alarm volume the user had before we raised it, and what we raised it to.
     * Persisted so a process death mid-alarm still puts their phone back the way they left it —
     * and the "raised to" half is what lets us tell our own change from one the user made since,
     * so we never silence the alarm clock they set in the meantime.
     */
    fun alarmVolumeRestore(): Pair<Int, Int>? {
        val previous = kv.get(KEY_ALARM_VOLUME_PREV)?.toIntOrNull() ?: return null
        val raisedTo = kv.get(KEY_ALARM_VOLUME_RAISED)?.toIntOrNull() ?: return null
        return previous to raisedTo
    }

    fun setAlarmVolumeRestore(previous: Int?, raisedTo: Int?) {
        if (previous == null || raisedTo == null) {
            kv.remove(KEY_ALARM_VOLUME_PREV)
            kv.remove(KEY_ALARM_VOLUME_RAISED)
        } else {
            kv.put(KEY_ALARM_VOLUME_PREV, previous.toString())
            kv.put(KEY_ALARM_VOLUME_RAISED, raisedTo.toString())
        }
    }

    /**
     * ALRM-17: the learned crying-alarm tuning, per camera. Plain steps; the bounds and their
     * meaning live in the monitor layer — corrupt or absent data reads as "nothing learned".
     */
    fun cryCalibrationSteps(did: String): Int = try {
        kv.get(KEY_CRY_CALIBRATION)?.let { JSONObject(it).optInt(did, 0) } ?: 0
    } catch (_: Exception) {
        0
    }

    fun saveCryCalibrationSteps(did: String, steps: Int) {
        val all = try {
            JSONObject(kv.get(KEY_CRY_CALIBRATION) ?: "{}")
        } catch (_: Exception) {
            JSONObject() // corrupt store: better to relearn than to fail to save (ALRM-17)
        }
        kv.put(KEY_CRY_CALIBRATION, all.put(did, steps).toString())
    }

    /** AUTH-10: signing out forgets the session AND the selected camera (not settings — and not
     *  the learned tuning, which belongs to the camera, not the account). */
    fun signOut() {
        kv.remove(KEY_SESSION)
        kv.remove(KEY_DEVICE)
    }
}

File: windows/src/BabyMonitor.Core/Data/Store.cs
using BabyMonitor.Core.Json;
using BabyMonitor.Core.Logging;
using BabyMonitor.Core.Xiaomi;

namespace BabyMonitor.Core.Data;

/// <summary>Where non-secret state lives — settings, the camera choice. The app supplies the file.</summary>
public interface IKeyValueStore
{
    string? Get(string key);

    void Put(string key, string value);

    void Remove(string key);
}

/// <summary>
/// Encryption seam for secrets — DPAPI on Windows, the Keychain on the Mac, the Keystore on the
/// phone, a passthrough fake in tests.
///
/// **Failure is a null, never an exception.** A secret store can refuse — a roamed profile, a
/// corrupted blob, a machine key rotated by an OS repair — and the caller must be able to answer
/// that by dropping the session (AUTH-6), not by dying. A value cannot be misused; an exception
/// crossing this seam once took a whole monitor down with it.
/// </summary>
public interface ISecretBox
{
    /// <summary>Null when the secret could not be sealed. The caller must then NOT persist it (AUTH-6).</summary>
    string? Seal(string plain);

    /// <summary>Null when the secret could not be read back — a lost session, never a crash.</summary>
    string? Open(string sealed_);
}

public static class Codecs
{
    public static string SessionToJson(Session s) => new JsonObj()
        .Put("userId", s.UserId)
        .Put("cUserId", s.CUserId)
        .Put("passToken", s.PassToken)
        .Put("serviceToken", s.ServiceToken)
        .Put("ssecurity", s.Ssecurity.ToBase64())
        .Put("region", s.Region)
        .ToString();

    public static Session? SessionFromJson(string json)
    {
        try
        {
            var v = new JsonObj(json);
            return new Session(
                UserId: v.GetString("userId"),
                CUserId: v.OptString("cUserId"),
                PassToken: v.GetString("passToken"),
                ServiceToken: v.GetString("serviceToken"),
                Ssecurity: v.GetString("ssecurity").Base64ToBytes(),
                Region: v.GetString("region"));
        }
        catch (Exception e) when (e is JsonException or FormatException or ArgumentException)
        {
            return null;
        }
    }

    public static string DeviceToJson(Device d) => new JsonObj()
        .Put("did", d.Did)
        .Put("name", d.Name)
        .Put("model", d.Model)
        .Put("mac", d.Mac)
        .Put("ip", d.Ip)
        .ToString();

    public static Device? DeviceFromJson(string json)
    {
        try
        {
            var v = new JsonObj(json);
            return new Device(
                Did: v.GetString("did"),
                Name: v.OptString("name"),
                Model: v.OptString("model"),
                Mac: v.OptString("mac"),
                Ip: v.OptString("ip"));
        }
        catch (Exception e) when (e is JsonException or FormatException)
        {
            return null;
        }
    }
}

/// <summary>
/// Everything the user can set. One shape, shared with the phone and the Mac — including its JSON,
/// its defaults, its clamps and its migrations (ALRM-6), so no platform can drift into meaning
/// something slightly different by "sensitivity 7".
/// </summary>
public sealed record Settings
{
    public const string ScheduleAlways = "always";
    public const string ScheduleWindow = "window";

    // ALRM-2: the sensitivity scale. 1 needs loud crying, 10 alarms on quiet crying.
    public const int SensitivityMin = 1;
    public const int SensitivityMax = 10;
    public const int SensitivityDefault = 5;

    // ALRM-4: never all the way down — a stored volume of zero would be a silent alarm.
    public const double VolumeMin = 0.2;
    public const double VolumeMax = 1.0;

    // WATCH-1: grace bounds — a zero grace would alarm on every one-second audio hiccup.
    public const int GraceMinSeconds = 5;
    public const int GraceMaxSeconds = 120;

    // ALRM-11: calm → urgent.
    public const string SoundSoftChime = "soft-chime";
    public const string SoundRisingChime = "rising-chime";
    public const string SoundLowPulse = "low-pulse";
    public const string SoundUrgentBeep = "urgent-beep";
    public const string SoundSiren = "siren";

    public static readonly IReadOnlyList<string> AlarmSounds = new[]
    {
        SoundSoftChime,
        SoundRisingChime,
        SoundLowPulse,
        SoundUrgentBeep,
        SoundSiren,
    };

    /// <summary>LIVE-18: the two pictures the camera offers. HD is what it actually has.</summary>
    public const string QualityHd = "hd";
    public const string QualitySd = "sd";

    public bool Muted { get; init; }

    /// <summary>
    /// LIVE-18: which picture this viewer asks the camera for. Unlike night vision (LIVE-10) it lives
    /// here rather than on the camera, because it is what *this* app requests — two people watching the
    /// same cot can choose differently. HD by default: the picture the camera has is the one to show
    /// unless someone says otherwise.
    /// </summary>
    public string VideoQuality { get; init; } = QualityHd;

    public bool AlarmEnabled { get; init; }

    /// <summary>ALRM-2: the one detection control — 1 (needs loud crying) .. 10 (quiet crying).</summary>
    public int AlarmSensitivity { get; init; } = SensitivityDefault;

    /// <summary>ALRM-7: "always" or a daily window (minutes since midnight, may cross midnight).</summary>
    public string AlarmScheduleMode { get; init; } = ScheduleAlways;

    public int AlarmWindowStartMinutes { get; init; } = 19 * 60;

    public int AlarmWindowEndMinutes { get; init; } = 7 * 60;

    /// <summary>WATCH-1: feed watchdog.</summary>
    public bool WatchdogEnabled { get; init; }

    public int WatchdogGraceSeconds { get; init; } = 30;

    // ALRM-11: how each alarm sounds. The default sounds differ so the two alarms are told apart by
    // ear out of the box (WATCH-2). A PC has no vibration motor, so the vibrate flags the phone
    // carries are read and written but never offered here (DESK-23).
    public string CryAlarmSound { get; init; } = SoundRisingChime;

    public string FeedAlarmSound { get; init; } = SoundLowPulse;

    public double CryAlarmVolume { get; init; } = 0.85;

    public double FeedAlarmVolume { get; init; } = 0.85;

    public bool CryAlarmVibrate { get; init; } = true;

    public bool FeedAlarmVibrate { get; init; } = true;

    public string ToJson() => new JsonObj()
        .Put("muted", Muted)
        .Put("videoQuality", VideoQuality)
        .Put("alarmEnabled", AlarmEnabled)
        .Put("alarmSensitivity", AlarmSensitivity)
        .Put("alarmScheduleMode", AlarmScheduleMode)
        .Put("alarmWindowStartMinutes", AlarmWindowStartMinutes)
        .Put("alarmWindowEndMinutes", AlarmWindowEndMinutes)
        .Put("watchdogEnabled", WatchdogEnabled)
        .Put("watchdogGraceSeconds", WatchdogGraceSeconds)
        .Put("cryAlarmSound", CryAlarmSound)
        .Put("feedAlarmSound", FeedAlarmSound)
        .Put("cryAlarmVolume", CryAlarmVolume)
        .Put("feedAlarmVolume", FeedAlarmVolume)
        .Put("cryAlarmVibrate", CryAlarmVibrate)
        .Put("feedAlarmVibrate", FeedAlarmVibrate)
        .ToString();

    public static Settings FromJson(string? json)
    {
        if (json == null)
        {
            return new Settings();
        }

        try
        {
            var v = new JsonObj(json);
            // ALRM-6: settings written before volume/vibrate were per-alarm had one shared
            // "alarmVolume"/"alarmVibrate" — an old value seeds both alarms rather than being lost.
            var legacyVolume = v.OptDouble("alarmVolume", 0.85);
            var legacyVibrate = v.OptBoolean("alarmVibrate", true);
            return new Settings
            {
                Muted = v.OptBoolean("muted", false),
                // Anything unrecognised is HD, not a stored string handed to the camera: a settings
                // file from a newer version, or a hand edit, must not end as a quality the camera does
                // not understand and a feed that never starts.
                VideoQuality = v.OptString("videoQuality", QualityHd) == QualitySd ? QualitySd : QualityHd,
                AlarmEnabled = v.OptBoolean("alarmEnabled", false),
                AlarmSensitivity = Math.Clamp(
                    v.OptInt("alarmSensitivity", LegacySensitivity(v)),
                    SensitivityMin,
                    SensitivityMax),
                AlarmScheduleMode = v.OptString("alarmScheduleMode", ScheduleAlways),
                AlarmWindowStartMinutes = Math.Clamp(v.OptInt("alarmWindowStartMinutes", 19 * 60), 0, (24 * 60) - 1),
                AlarmWindowEndMinutes = Math.Clamp(v.OptInt("alarmWindowEndMinutes", 7 * 60), 0, (24 * 60) - 1),
                WatchdogEnabled = v.OptBoolean("watchdogEnabled", false),
                // Clamped like sensitivity: stored data (old versions, hand edits) must never yield a
                // watchdog that fires on every hiccup or an alarm that plays silently.
                WatchdogGraceSeconds = Math.Clamp(
                    v.OptInt("watchdogGraceSeconds", 30),
                    GraceMinSeconds,
                    GraceMaxSeconds),
                CryAlarmSound = v.OptString("cryAlarmSound", SoundRisingChime),
                FeedAlarmSound = v.OptString("feedAlarmSound", SoundLowPulse),
                CryAlarmVolume = Math.Clamp(v.OptDouble("cryAlarmVolume", legacyVolume), VolumeMin, VolumeMax),
                FeedAlarmVolume = Math.Clamp(v.OptDouble("feedAlarmVolume", legacyVolume), VolumeMin, VolumeMax),
                CryAlarmVibrate = v.OptBoolean("cryAlarmVibrate", legacyVibrate),
                FeedAlarmVibrate = v.OptBoolean("feedAlarmVibrate", legacyVibrate),
            };
        }
        catch (JsonException)
        {
            return new Settings();
        }
    }

    /// <summary>
    /// ALRM-6: settings written before the sensitivity slider carried a dB threshold (3 = most
    /// sensitive .. 24 = least). Map it onto the 1..10 scale rather than lose the user's tuning. The
    /// old scale is frozen history — these constants must never track whatever the live sensitivity
    /// mapping becomes.
    /// </summary>
    private static int LegacySensitivity(JsonObj v)
    {
        if (!v.Has("alarmThresholdDb"))
        {
            return SensitivityDefault;
        }

        var mapped = (int)Math.Round((24.0 - v.OptDouble("alarmThresholdDb", 14.0)) / 2.0, MidpointRounding.AwayFromZero);
        return Math.Clamp(mapped, SensitivityMin, SensitivityMax);
    }
}

public sealed class AppStore
{
    private const string KeySession = "session_v1";
    private const string KeyDevice = "device_v1";
    private const string KeySettings = "settings_v1";
    private const string KeyMonitoring = "monitoring_v1";
    private const string KeyCryCalibration = "cry_calibration_v1"; // JSON object: camera did → steps

    private readonly IKeyValueStore _kv;
    private readonly ISecretBox _box;

    /// <summary>
    /// The decrypted session, held after the first read. Every caller that wants to know whether we
    /// are signed in — the router, the engine, the camera controls — would otherwise pay a full
    /// decrypt, and the UI asks on every level-meter tick.
    /// </summary>
    private volatile Session? _cachedSession;

    private volatile bool _sessionRead;

    public AppStore(IKeyValueStore kv, ISecretBox box)
    {
        _kv = kv;
        _box = box;
    }

    /// <summary>
    /// AUTH-6: tokens are only ever stored encrypted. If the secret store refuses, drop the session
    /// rather than crash or, worse, fall back to plaintext — the user simply signs in again.
    /// </summary>
    public bool SaveSession(Session s)
    {
        string? sealed_;
        try
        {
            sealed_ = _box.Seal(Codecs.SessionToJson(s));
        }
        catch (Exception e)
        {
            Log.W("data", $"could not encrypt session: {e.Message}", e);
            sealed_ = null;
        }

        if (sealed_ == null)
        {
            Log.W("data", "the secret store refused the session — not persisting it");
            _kv.Remove(KeySession);
            _cachedSession = null;
            _sessionRead = true;
            // AUTH-13: report the failure so a shell can say so, never a silent "signed in" that is
            // then quietly dropped back to the login screen.
            return false;
        }

        _kv.Put(KeySession, sealed_);
        _cachedSession = s;
        _sessionRead = true;
        return true;
    }

    public Session? LoadSession()
    {
        if (_sessionRead)
        {
            return _cachedSession;
        }

        var stored = _kv.Get(KeySession);
        var plain = stored == null ? null : _box.Open(stored);
        _cachedSession = plain == null ? null : Codecs.SessionFromJson(plain);
        _sessionRead = true;
        return _cachedSession;
    }

    public void SaveDevice(Device d) => _kv.Put(KeyDevice, Codecs.DeviceToJson(d));

    public Device? LoadDevice()
    {
        var stored = _kv.Get(KeyDevice);
        return stored == null ? null : Codecs.DeviceFromJson(stored);
    }

    public Settings LoadSettings() => Settings.FromJson(_kv.Get(KeySettings));

    public void SaveSettings(Settings s) => _kv.Put(KeySettings, s.ToJson());

    /// <summary>BG-13: was monitoring running when the process last went away (a restart, a crash)?</summary>
    public bool WasMonitoring() => _kv.Get(KeyMonitoring) == "true";

    public void SetMonitoring(bool on) => _kv.Put(KeyMonitoring, on ? "true" : "false");

    /// <summary>
    /// ALRM-17: the learned crying-alarm tuning, per camera. Plain steps; the bounds and their
    /// meaning live in the monitor layer — corrupt or absent data reads as "nothing learned".
    /// </summary>
    public int CryCalibrationSteps(string did)
    {
        try
        {
            var stored = _kv.Get(KeyCryCalibration);
            return stored == null ? 0 : new JsonObj(stored).OptInt(did, 0);
        }
        catch (JsonException)
        {
            return 0;
        }
    }

    public void SaveCryCalibrationSteps(string did, int steps)
    {
        JsonObj all;
        try
        {
            all = new JsonObj(_kv.Get(KeyCryCalibration) ?? "{}");
        }
        catch (JsonException)
        {
            all = new JsonObj(); // corrupt store: better to relearn than to fail to save (ALRM-17)
        }

        _kv.Put(KeyCryCalibration, all.Put(did, steps).ToString());
    }

    /// <summary>
    /// AUTH-10: signing out forgets the session AND the selected camera (not settings — and not the
    /// learned tuning, which belongs to the camera, not the account).
    /// </summary>
    public void SignOut()
    {
        _kv.Remove(KeySession);
        _kv.Remove(KeyDevice);
        _cachedSession = null;
        _sessionRead = true;
    }

    /// <summary>CAM-4: switching camera forgets the choice, not the account — the picker comes back.</summary>
    public void ClearDevice() => _kv.Remove(KeyDevice);
}
